api/views/company.py

```diff
@@ -140,14 +146,55 @@ class CompanyCreateView(CreateAPIView):
     @transaction.atomic
     def perform_create(self, serializer):
         new_company = serializer.save()
-        supervisor, _ = CompanySupervisor.objects.get_or_create(user=self.request.user)
-        supervisor.companies.add(new_company)  # NOTE: on suppose que `new_company` n'est pas déjà dedans
+        SupervisorRole.objects.create(company=new_company, user=self.request.user)
         return new_company
 
 
 class CompanyRetrieveView(RetrieveAPIView):
     """Récupération d'une entreprise dont l'utilisateur est gestionnaire"""
 
     queryset = Company.objects.all()
-    permission_classes = [IsSupervisorOfThisCompany]
+    permission_classes = [IsSupervisor]
     serializer_class = CompanySerializer
+
+
+class GetCompanyCollaboratorsView(APIView):
```

Comment:
Ce serait mieux d'hériter de ListApiView car : 
- En spécifiant le context manuellement comme dans 166, on n'envoie pas les autres éléments du contexte que DRF envoie automatiquement aussi (la _view_, le _format_, et la _request_)
- On n'aura pas forcément besoin d'envoyer le `company_id` dans le contexte du serializer (même si on peut si on veut un peu de sugar)
- Plus cohérent avec les autres vues liste de l'application

On perd les autres éléments de context en le mettant directement (DRF envoie aussi la view, le format et la request, d'ailleurs on n'a en vrai même pas besoin d'envoyer le pk de l'entreprise en context)

```python
class GetCompanyCollaboratorsView(ListAPIView):
    """Récupération des utilisateurs ayant au moins un rôle dans cette entreprise"""

    model = get_user_model()
    serializer_class = CollaboratorSerializer

    def get_queryset(self):
        company = get_object_or_404(Company.objects.filter(supervisors=self.request.user), pk=self.kwargs[self.lookup_field])
        return company.collaborators.all()

    # Vu que par défault on a la view dans le contexte du serializer, on n'a p-e pas besoin d'envoyer ce contexte, mais
    # bon ça fait aussi un peu de sugar, ça ne me dérange pas de le garder non plus
    def get_serializer_context(self):
        context = super().get_serializer_context()
        context.update({"company_id": self.kwargs[self.kwargs[self.lookup_field]]})
        return context
```

Reply:
Merci ! Je reprends tes points pour être sûr de bien comprendre et modifier en conséquence dans le bon sens ensuite : 

> En spécifiant le context manuellement comme dans 166, on n'envoie pas les autres éléments du contexte que DRF envoie automatiquement aussi (la view, le format, et la request)

Oui en quoi est-ce important si là en l'occurence on n'en a pas besoin ?

> On n'aura pas forcément besoin d'envoyer le company_id dans le contexte du serializer (même si on peut si on veut un peu de sugar)

J'ai pas bien compris cette partie. Quelle est l'alternative sans passer `company_id` dans le serializer? Est-ce que tu as un exemple de code pour illustrer please ?

> Plus cohérent avec les autres vues liste de l'application

En gros, si j'ai bien compris, dans le code de manière générale, tu préfères avoir uniquement des vues génériques par souci d'uniformité (sauf si c'est vraiment pas possible) ?

Moi j'avais plutôt une approche différente sur le sujet, en mode "right tool for the task". En gros quand c'est CRUD ou presque, plutôt vue générique, sinon plutôt vue classique. Parce que si j'override tous les comportements par défaut parce qu'ils sont pas les bons, c'est comme si la vue générique ne servait à rien, et le code final produit est plus complexe.

Mais si tu veux qu'on parte sur du générique par défaut, pas de souci je m'adapte !